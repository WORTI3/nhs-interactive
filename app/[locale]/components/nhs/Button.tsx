import React, { ReactNode } from 'react';

interface ButtonProps {
  id?: string;
  className?: string;
  type?: "submit" | "button" | "reset";
  secondary?: boolean;
  reverse?: boolean;
  preventDoubleClick?: boolean;
  as?: 'button' | 'a';
  href?: string;
  onClick?: (event: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void;
  disabled?: boolean;
  children: ReactNode;
}

/**
 * Represents a customizable button component.
 *
 * @component
 * @param {object} props - The properties of the component.
 * @param {string} [props.id] - The unique identifier for the button.
 * @param {string} [props.className] - Additional class names to be applied to the button.
 * @param {"submit" | "button" | "reset"} [props.type="button"] - The type of the button (submit, button, reset).
 * @param {boolean} [props.secondary=false] - Determines if the button is of secondary style.
 * @param {boolean} [props.reverse=false] - Determines if the button is of reverse style.
 * @param {boolean} [props.preventDoubleClick=false] - Determines if double-click prevention should be applied.
 * @param {'button' | 'a'} [props.as='button'] - The type of element to render (button or anchor).
 * @param {string} [props.href] - The URL to navigate to if the button is rendered as an anchor.
 * @param {(event: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void} [props.onClick] - The callback function for the click event.
 * @param {boolean} [props.disabled] - Determines if the button is disabled.
 * @param {ReactNode} props.children - The content of the button.
 * @returns {ReactNode} The rendered button component.
 */
const Button: React.FC<ButtonProps> = ({
  id,
  className,
  type = "button",
  secondary = false,
  reverse = false,
  preventDoubleClick = false,
  as = 'button',
  href,
  onClick,
  disabled,
  children
}) => {
  const buttonClasses = `nhsuk-button ${secondary ? 'nhsuk-button--secondary' : ''} ${reverse ? 'nhsuk-button--reverse' : ''} ${className || ''}`;

  const handleClick = (event: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => {
    if (preventDoubleClick) {
      // @ts-ignore
      event.currentTarget.disabled = true;
    }
    if (onClick) {
      onClick(event);
    }
  };

  if (as === 'a') {
    return (
      <a
        id={id}
        className={buttonClasses}
        href={href}
        onClick={handleClick}
        // @ts-ignore
        disabled={disabled}
      >
        {children}
      </a>
    );
  }

  return (
    <button
      id={id}
      className={buttonClasses}
      type={type}
      data-module="nhsuk-button"
      data-prevent-double-click={preventDoubleClick}
      onClick={handleClick}
      disabled={disabled}
    >
      {children}
    </button>
  );
};

export default Button;
