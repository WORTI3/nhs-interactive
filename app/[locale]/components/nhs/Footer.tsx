"use client";

import { Footer as NHSFooter } from "nhsuk-react-components";

export const Footer = () => {
  return (
    <NHSFooter>
      <NHSFooter.List>
        <NHSFooter.ListItem href="https://github.com/worti3" target="_blank" rel="noreferrer" aria-label="Github Profile">Github</NHSFooter.ListItem>
      </NHSFooter.List>
      {/* <NHSFooter.Copyright>&copy; </NHSFooter.Copyright> */}
    </NHSFooter>
  );
}
