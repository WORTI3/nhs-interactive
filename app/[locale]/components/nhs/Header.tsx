"use client";

import { Header as NHSHeader } from "nhsuk-react-components";
import LanguageSelector from "../LanguageSelector";
import { useTranslations } from "next-intl";

export const Header = ({locale}: {locale: string}) => {
  const translate = useTranslations('header');
  return (
    <>
      <NHSHeader transactional>
        <NHSHeader.Container>
          <NHSHeader.Logo href="/" />
          <NHSHeader.ServiceName href="/">{translate('title')}</NHSHeader.ServiceName>
          <LanguageSelector translate={translate} locale={locale} />
        </NHSHeader.Container>
      </NHSHeader>
    </>
  );
}
