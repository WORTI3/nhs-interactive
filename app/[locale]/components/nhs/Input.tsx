import { classNames } from '../../../utils';
import { parseAsString, useQueryState } from 'next-usequerystate';
import React from 'react';

interface InputProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  id?: string;
  label?: string;
  hint?: string;
  error?: string | null;
}

/**
 * Represents a customisable input component.
 *
 * @component
 * @param {object} props - The properties of the component.
 * @param {string} [props.id] - The unique identifier for the input.
 * @param {string} [props.label] - The label for the input.
 * @param {string} [props.hint] - The hint for the input.
 * @param {string | null} [props.error] - The error message for the input.
 * @param {ReactNode} props.children - The content of the input.
 * @returns {ReactNode} The rendered input component.
 */
const Input: React.FC<InputProps> = ({
  id,
  label,
  hint,
  error,
  ...restProps
}) => {
  const [rows, setRows] = React.useState(1);
  const [query, setQuery] = useQueryState('q', parseAsString.withDefault(""));
  const [showText, setShowText] = React.useState(false);
  const handleChange = (event: { target: { value: React.SetStateAction<string>; }; }) => {
    if (event.target.value.length >= 50 && query.length >= 50) {
      setShowText(true);
      setRows(3);
    } else {
      setShowText(false);
      // dont set rows here
    }
    
    setQuery(event.target.value);
  };

  const inputContainerClasses = `nhsuk-character-count nhsuk-form-group ${error ? 'nhsuk-form-group--error' : ''}`;
  const textBoxClasses = `nhsuk-textarea nhsuk-js-character-count ${error ? 'nhsuk-textarea--error' : ''}`;
  
  return (
    <div className={inputContainerClasses}>
      {label && <label className="nhsuk-label" htmlFor={id}>{label}</label>}
      {hint && <span className="nhsuk-hint">{hint}</span>}
      {error && <span className="nhsuk-error-message">{error}</span>}
      <textarea 
        id={id}
        value={query}
        className={textBoxClasses} 
        rows={rows}
        maxLength={100}
        onChange={handleChange} 
        {...restProps}></textarea>
    
      { showText &&
        <div className={classNames((query.length > 95) ? "nhsuk-error-message" : "nhsuk-hint")} id="exceeding-info">
          You have {100 - query.length} characters remaining
        </div>
      }
    </div>
  );
};

export default Input;
