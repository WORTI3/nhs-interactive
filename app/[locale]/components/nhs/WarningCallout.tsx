import React, { ReactNode, ChangeEvent } from 'react';

interface WarningCalloutProps {
  className?: string;
  children: ReactNode;
}

/**
 * Represents a warning callout component.
 *
 * @component
 * @param {object} props - The properties of the component.
 * @param {string} [props.className] - Additional class names to be applied to the warning callout.
 * @param {ReactNode} props.children - The content of the warning callout.
 * @returns {ReactNode} The rendered warning callout component.
 */
export const WarningCallout: React.FC<WarningCalloutProps> = ({ className, children }) => {
  const containerClasses = `${className || ''} nhsuk-warning-callout`;

  return (
    <div className={containerClasses}>
      {children}
    </div>
  );
};

interface WarningCalloutLabelProps {
  children: ReactNode;
}

/**
 * Represents a label within a warning callout.
 *
 * @component
 * @param {object} props - The properties of the component.
 * @param {ReactNode} props.children - The content of the label.
 * @returns {ReactNode} The rendered label component.
 */
export const WarningCalloutLabel: React.FC<WarningCalloutLabelProps> = ({ children }) => (
  <h2 className="nhsuk-warning-callout__label">
    <span role="text">
      <span className="nhsuk-u-visually-hidden">Important: </span>
      {children}
    </span>
  </h2>
);

interface WarningCalloutTextProps {
  children: ReactNode;
}

/**
 * Can be removed in future.. relays children given in a p tag.
 * 
 * @component
 * @param {object} props - The properties of the component.
 * @returns {ReactNode} The rendered text component.
 */
export const WarningCalloutText: React.FC<WarningCalloutTextProps> = ({ children }) => {
  return <p>{children}</p>;
};
