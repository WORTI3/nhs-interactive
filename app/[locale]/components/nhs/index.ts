export { default as Button } from './Button';
export { default as Input } from './Input';
export { Header } from './Header';
export { Footer } from './Footer';
export { wrapper as Wrapper } from './wrapper';
export * from './WarningCallout';