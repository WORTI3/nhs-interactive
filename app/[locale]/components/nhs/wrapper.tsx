// not a true react component
export function wrapper(content: any) {
  return (
    <div className="nhsuk-width-container">
      <main className="nhsuk-main-wrapper" id="maincontent" role="main">
        <div className="nhsuk-grid-row">
          <div className="nhsuk-grid-column-two-thirds">
            { content }
          </div>
        </div>
      </main>
    </div>
  );
}
