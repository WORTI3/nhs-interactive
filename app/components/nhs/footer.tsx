"use client";

import { Footer as NHSFooter } from "nhsuk-react-components";

export const Footer = () => {
  return (
    <NHSFooter>
      <NHSFooter.List>
        <NHSFooter.ListItem href="/">back to top</NHSFooter.ListItem>
      </NHSFooter.List>
      <NHSFooter.Copyright>&copy; </NHSFooter.Copyright>
    </NHSFooter>
  );
}