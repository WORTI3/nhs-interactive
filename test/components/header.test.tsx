import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Header } from '../../app/[locale]/components/nhs/Header';
import { load } from 'cheerio';

describe('NHSHeader component', () => {
  test('rendered component should match snapshot', () => {
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });

  test('should render the component with the correct links & elements', () => {
    const { container } = render(<Header />);
    const html = container.innerHTML;
    const $ = load(html);
    const nhsHeader = $('.nhsuk-header');
  
    expect(nhsHeader.text()).toStrictEqual('NHS LogoInteractive');
    expect(nhsHeader.find('p').length).toBe(0);
    const headerLinks = nhsHeader.find('a');
    expect(headerLinks.length).toBe(2);
    expect(headerLinks.text()).toStrictEqual('NHS LogoInteractive');
  });
});
