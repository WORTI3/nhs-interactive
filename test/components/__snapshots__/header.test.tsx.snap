// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NHSHeader component rendered component should match snapshot (cy) 1`] = `
<div>
  <header
    class="nhsuk-header nhsuk-header--transactional"
    role="banner"
  >
    <div
      class="nhsuk-width-container nhsuk-header__container"
    >
      <div
        class="nhsuk-header__logo nhsuk-header__logo--only"
      >
        <a
          aria-label="NHS homepage"
          class="nhsuk-header__link"
          href="/"
        >
          <svg
            aria-labelledby="nhsuk-logo_title"
            class="nhsuk-logo"
            focusable="false"
            role="img"
            viewBox="0 0 40 16"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <title
              id="nhsuk-logo_title"
            >
              NHS Logo
            </title>
            <path
              class="nhsuk-logo__background"
              d="M0 0h40v16H0z"
            />
            <path
              class="nhsuk-logo__text"
              d="M3.9 1.5h4.4l2.6 9h.1l1.8-9h3.3l-2.8 13H9l-2.7-9h-.1l-1.8 9H1.1M17.3 1.5h3.6l-1 4.9h4L25 1.5h3.5l-2.7 13h-3.5l1.1-5.6h-4.1l-1.2 5.6h-3.4M37.7 4.4c-.7-.3-1.6-.6-2.9-.6-1.4 0-2.5.2-2.5 1.3 0 1.8 5.1 1.2 5.1 5.1 0 3.6-3.3 4.5-6.4 4.5-1.3 0-2.9-.3-4-.7l.8-2.7c.7.4 2.1.7 3.2.7s2.8-.2 2.8-1.5c0-2.1-5.1-1.3-5.1-5 0-3.4 2.9-4.4 5.8-4.4 1.6 0 3.1.2 4 .6"
            />
            <image
              src="https://assets.nhs.uk/images/nhs-logo.png"
              xlink:href=""
            />
          </svg>
        </a>
      </div>
      <div
        class="nhsuk-header__transactional-service-name"
      >
        <a
          class="nhsuk-header__transactional-service-name--link"
          href="/"
        >
          Rhyngweithiol
        </a>
      </div>
      <div
        class="relative max-lg:absolute max-lg:right-4 float-right inline-block text-left"
        data-headlessui-state=""
      >
        <div>
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            class="inline-flex w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
            data-headlessui-state=""
            id="language-button"
            type="button"
          >
            Iaith
            <svg
              aria-hidden="true"
              class="-mr-1 h-5 w-5 text-gray-400"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
              />
              <path
                d="M3.6 9l16.8 0"
              />
              <path
                d="M3.6 15l16.8 0"
              />
              <path
                d="M11.5 3a17 17 0 0 0 0 18"
              />
              <path
                d="M12.5 3a17 17 0 0 1 0 18"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`NHSHeader component rendered component should match snapshot (en) 1`] = `
<div>
  <header
    class="nhsuk-header nhsuk-header--transactional"
    role="banner"
  >
    <div
      class="nhsuk-width-container nhsuk-header__container"
    >
      <div
        class="nhsuk-header__logo nhsuk-header__logo--only"
      >
        <a
          aria-label="NHS homepage"
          class="nhsuk-header__link"
          href="/"
        >
          <svg
            aria-labelledby="nhsuk-logo_title"
            class="nhsuk-logo"
            focusable="false"
            role="img"
            viewBox="0 0 40 16"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <title
              id="nhsuk-logo_title"
            >
              NHS Logo
            </title>
            <path
              class="nhsuk-logo__background"
              d="M0 0h40v16H0z"
            />
            <path
              class="nhsuk-logo__text"
              d="M3.9 1.5h4.4l2.6 9h.1l1.8-9h3.3l-2.8 13H9l-2.7-9h-.1l-1.8 9H1.1M17.3 1.5h3.6l-1 4.9h4L25 1.5h3.5l-2.7 13h-3.5l1.1-5.6h-4.1l-1.2 5.6h-3.4M37.7 4.4c-.7-.3-1.6-.6-2.9-.6-1.4 0-2.5.2-2.5 1.3 0 1.8 5.1 1.2 5.1 5.1 0 3.6-3.3 4.5-6.4 4.5-1.3 0-2.9-.3-4-.7l.8-2.7c.7.4 2.1.7 3.2.7s2.8-.2 2.8-1.5c0-2.1-5.1-1.3-5.1-5 0-3.4 2.9-4.4 5.8-4.4 1.6 0 3.1.2 4 .6"
            />
            <image
              src="https://assets.nhs.uk/images/nhs-logo.png"
              xlink:href=""
            />
          </svg>
        </a>
      </div>
      <div
        class="nhsuk-header__transactional-service-name"
      >
        <a
          class="nhsuk-header__transactional-service-name--link"
          href="/"
        >
          Interactive
        </a>
      </div>
      <div
        class="relative max-lg:absolute max-lg:right-4 float-right inline-block text-left"
        data-headlessui-state=""
      >
        <div>
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            class="inline-flex w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
            data-headlessui-state=""
            id="language-button"
            type="button"
          >
            Language
            <svg
              aria-hidden="true"
              class="-mr-1 h-5 w-5 text-gray-400"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
              />
              <path
                d="M3.6 9l16.8 0"
              />
              <path
                d="M3.6 15l16.8 0"
              />
              <path
                d="M11.5 3a17 17 0 0 0 0 18"
              />
              <path
                d="M12.5 3a17 17 0 0 1 0 18"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`NHSHeader component rendered component should match snapshot (gd) 1`] = `
<div>
  <header
    class="nhsuk-header nhsuk-header--transactional"
    role="banner"
  >
    <div
      class="nhsuk-width-container nhsuk-header__container"
    >
      <div
        class="nhsuk-header__logo nhsuk-header__logo--only"
      >
        <a
          aria-label="NHS homepage"
          class="nhsuk-header__link"
          href="/"
        >
          <svg
            aria-labelledby="nhsuk-logo_title"
            class="nhsuk-logo"
            focusable="false"
            role="img"
            viewBox="0 0 40 16"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <title
              id="nhsuk-logo_title"
            >
              NHS Logo
            </title>
            <path
              class="nhsuk-logo__background"
              d="M0 0h40v16H0z"
            />
            <path
              class="nhsuk-logo__text"
              d="M3.9 1.5h4.4l2.6 9h.1l1.8-9h3.3l-2.8 13H9l-2.7-9h-.1l-1.8 9H1.1M17.3 1.5h3.6l-1 4.9h4L25 1.5h3.5l-2.7 13h-3.5l1.1-5.6h-4.1l-1.2 5.6h-3.4M37.7 4.4c-.7-.3-1.6-.6-2.9-.6-1.4 0-2.5.2-2.5 1.3 0 1.8 5.1 1.2 5.1 5.1 0 3.6-3.3 4.5-6.4 4.5-1.3 0-2.9-.3-4-.7l.8-2.7c.7.4 2.1.7 3.2.7s2.8-.2 2.8-1.5c0-2.1-5.1-1.3-5.1-5 0-3.4 2.9-4.4 5.8-4.4 1.6 0 3.1.2 4 .6"
            />
            <image
              src="https://assets.nhs.uk/images/nhs-logo.png"
              xlink:href=""
            />
          </svg>
        </a>
      </div>
      <div
        class="nhsuk-header__transactional-service-name"
      >
        <a
          class="nhsuk-header__transactional-service-name--link"
          href="/"
        >
          Ìre-thaiceil
        </a>
      </div>
      <div
        class="relative max-lg:absolute max-lg:right-4 float-right inline-block text-left"
        data-headlessui-state=""
      >
        <div>
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            class="inline-flex w-full justify-center gap-x-1.5 rounded-md bg-white px-3 py-2 text-sm font-semibold text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 hover:bg-gray-50"
            data-headlessui-state=""
            id="language-button"
            type="button"
          >
            Cànan
            <svg
              aria-hidden="true"
              class="-mr-1 h-5 w-5 text-gray-400"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
              />
              <path
                d="M3.6 9l16.8 0"
              />
              <path
                d="M3.6 15l16.8 0"
              />
              <path
                d="M11.5 3a17 17 0 0 0 0 18"
              />
              <path
                d="M12.5 3a17 17 0 0 1 0 18"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;
